<a href="https://colab.research.google.com/github/Anaya666/project-1-DS1002-/blob/main/Assignments/14_TextGeneration.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Final homework assignment:  generating text from a corpus

Due Friday 3 May at 5 pm

We begin by loading all of Shakespeare's plays into a text file, and separating them into individual words.  This is made easy by the fact that someone already made a giant text file of these plays for us, and put it online.

[Note that by changing the url in the following code chunk, you can load any text file that lives online in this way . . . see below.]

Run the following code.  It should output a list of ten words, starting with 'From' and ending with 'rose'.

In [1]:
from urllib.request import urlopen            #package for reading info from the web
url='https://ocw.mit.edu/ans7870/6/6.006/s08/lecturenotes/files/t8.shakespeare.txt'   #web address
text = urlopen(url).read().decode('utf-8')    #open web address, read file, convert to text
shakespeare_words=text.split()[1651:]         #split text into words (not including preamble)
shakespeare_words[:10]                        #display the first 10 words

['From',
 'fairest',
 'creatures',
 'we',
 'desire',
 'increase,',
 'That',
 'thereby',
 "beauty's",
 'rose']

## 1. `nextwords` practice

Create a practice function called `createNextwordsPractice` which takes a list of words called `wordlist` and returns a dictionary called `nextwordsPractice`. The keys of `nextwordsPractice` should be all the words in `wordlist`, and the value for any given key should be a list of all the words that follow the key word somewhere in the original `wordlist`.



In [3]:
def createNextwordsPractice(wordlist): #this is an empty dictionary that will store our results
  nextwordsPractice = {}
  for i in range(len(wordlist)): #looping through all words in wordlist
        word = wordlist[i] #For each index, we extract the corresponding word from 'wordlist' and assign it to the variable 'word'.
        if word not in nextwordsPractice:
            nextwordsPractice[word] = [] #If word not in nextwordsPractice, we added it as a new key to the dictionary and assigned an empty list as its value.

        if i < len(wordlist) - 1:
            next_word = wordlist[i + 1]
            nextwordsPractice[word].append(next_word) #checking if there is a word after the current word and retrieving the next word and adding it to the list

  return nextwordsPractice

Run the following code:

In [4]:
practice = createNextwordsPractice(shakespeare_words)
practice['From']

['fairest',
 'his',
 'sullen',
 'limits',
 'hands',
 'whence',
 'where',
 'me',
 'this',
 'thy',
 'hence',
 'thee,',
 'you',
 'heaven',
 'her',
 'simple',
 'whence',
 'below',
 'lowest',
 'courtly',
 'son',
 'the',
 'the',
 'Sicyon,',
 'Alexandria',
 'Silvius,',
 'mine',
 'the',
 'firm',
 'Egypt',
 'Antony',
 'thine',
 'that',
 'my',
 'me',
 'henceforth',
 'tyrant',
 'seventeen',
 'miserable',
 'whom',
 'my',
 'whence,',
 'me',
 'me',
 'slaves',
 'every',
 'rascals',
 'whom',
 'where',
 "th'",
 'Rome',
 'off',
 'these',
 'twelve',
 "th'",
 'whence?',
 'Rome.',
 'thy',
 'Gallia',
 'fairies',
 'off',
 "th'",
 'most',
 'thy',
 'youth',
 'every',
 'whose',
 'so',
 'one',
 'this',
 'the',
 'east',
 'action',
 'my',
 'this',
 'Leonati',
 'her',
 'chance',
 'south',
 'our',
 'whence',
 'the',
 'top',
 'this',
 'our',
 'that',
 'me,',
 'fashion',
 'the',
 'another',
 'Hamlet?',
 'whence',
 'whom',
 "Rumour's",
 'whence',
 'the',
 'a',
 'a',
 'enemies',
 'every',
 'this',
 'the',
 'open',
 'who

In [5]:
practice['rose']

['might',
 'looks',
 'is',
 'of',
 'Of',
 'may',
 'will',
 'of',
 'From',
 'and',
 'of',
 'both',
 'with',
 'from',
 'with',
 'with',
 'side.',
 'red,',
 'too.',
 'in',
 'a',
 'a',
 'I',
 'against',
 'that',
 'and',
 'from',
 'again,',
 'Lest',
 'against',
 'Than',
 "distill'd",
 'on',
 'up',
 'in',
 'wither.',
 'and',
 'By',
 'again?',
 'he']

In [6]:
practice['rose.']

['It']

## 2. `nextwords` for real:  new sentences.

Next, write a new function called `createNextwords` which takes a `wordlist` and returns a dictionary of `nextwords`, but now with a new additional feature.  
In addition to the keys from problem 1, add a special key to the dictionary, which will be the symbol '%'.  This key will signify the first word of a sentence. When scanning the wordlist to construct the `nextwords` dictionary, normally you would add `wordlist[i+1]` to the list `nextwords[wordlist[i]]`. However, if `wordlist[i]` ends with a period, then you should instead add `wordlist[i+1]` to the list nextwords['%']. Also add the very first word of `wordlist` to nextwords['%'].


You should be able to produce this output:
```
> shakespeare_nextwords=createNextwords(shakespeare_words)
> shakespeare_nextwords['Rose']
['at', 'or', 'or', 'and']

> shakespeare_nextwords['rose.']
[]

> len(shakespeare_nextwords['%'])
76543
```

In [18]:
def createNextwords(wordlist):
    nextwords = {}  # Initialize an empty dictionary to store the next words

    # Handling the first word of the text indicated by its index at 0
    nextwords['%'] = [wordlist[0]]

    # Iterating through the word list to construct the nextwords dictionary
    i = 0
    while i < len(wordlist) - 1:
        word = wordlist[i]
        next_word = wordlist[i + 1]

        # Check if the current word ends with a period and the next word starts with a capital letter
        if word.endswith('.') and next_word[0].isupper():
            # Add the next word to the list associated with the special key '%'
            if '%' not in nextwords:
                nextwords['%'] = [next_word]
            else:
                nextwords['%'].append(next_word)

            # Moving to the word after the next word
            i += 2
            continue

        # Check if the current word ends with a period
        elif word.endswith('.'):
            # If yes, add the next word to the list associated with the special key '%'
            if '%' not in nextwords:
                nextwords['%'] = [next_word]
            else:
                nextwords['%'].append(next_word)

        else:
            # If not, add the next word to the list associated with the current word
            if word not in nextwords:
                nextwords[word] = [next_word]
            else:
                nextwords[word].append(next_word)

        i += 1

    return nextwords

In [ ]:
#sources: https://docs.python.org/3/tutorial/datastructures.html
#https://www.geeksforgeeks.org/isupper-islower-lower-upper-python-applications/

Run the following code:

In [15]:
shakespeare_nextwords = createNextwords(shakespeare_words)

print(shakespeare_nextwords['Rose'])
print(len(shakespeare_nextwords['%']))

['at', 'or', 'or', 'and']
58493


In [16]:
print(len(shakespeare_nextwords['rose']))

40


In [22]:
print(shakespeare_nextwords['rose.'])

KeyError: 'rose.'

In [13]:
if 'rose.' in shakespeare_words:
  print("available")
else:
  print("The word 'rose.' does not appear in the text ")



available


## 3.  Make your own Shakespeare play

Okay, that may be a bit of an exaggeration.  Anyway, next up is text generation.  
We'll need the choice function:

In [23]:
from random import choice

This function chooses a random element from a list.  Run this code a few times:

In [28]:
choice([1, 17, 34, 289])

289

Now, create a function called `makePlay(nextwords,n)` which takes a dictionary `nextwords` and a number `n`, and follows these steps:
1. Define an empty `newplay` string.  (I did this for you.)
2. Set `currentword` equal to '%'.
3. Start a for-loop that will run `n` times.
4. Randomly choose a word from the list `nextwords[currentword]`.
5. Add the chosen word to `newplay`, followed by a space `' '` character.
6. If the chosen word ends in a period, add a newline character '\n' and set `currentword` equal to `%`.
7. Otherwise, set `currentword` equal to the chosen word.
8. Go back to step 4.
9. When all `n` iterations have finished, your function should return `newplay`.

In [32]:
def makePlay(nextwords, n):
    newplay = ''  # Initialize an empty string to store the generated text
    currentword = '%'  # Start with '%' as the current word

    # Start a for-loop that will run n times
    for _ in range(n):
        # Check if the current word exists in the nextwords dictionary- I added this additional step as there was an error that "Nathaniel's" is not present in the nextwords dictionary.
        if currentword in nextwords:
            # Randomly choose a word from the list nextwords[currentword]
            chosen_word = choice(nextwords[currentword])
        else:
            # If the current word does not exist in the dictionary, choose a new starting word
            chosen_word = choice(nextwords['%'])

        # Add the chosen word to newplay, followed by a space ' ' character
        newplay += chosen_word + ' '

        # If the chosen word ends in a period, add a newline character '\n' and set currentword equal to '%'
        if chosen_word.endswith('.'):
            newplay += '\n'
            currentword = '%'
        else:
            # Otherwise, set currentword equal to the chosen word
            currentword = chosen_word

    return newplay


Now create a 1000 word Shakesperian play by running the following code:

In [33]:
nextwords=createNextwords(shakespeare_words)
newplay=makePlay(nextwords,1000)
print(newplay)

Margaret may your daughters profit you hag, you. 
ANTONIO. 
FALSTAFF. 
If I think him the madman is; you must be his weakness, married to CLARENCE] Nay, then I'll still at our well-found successes as she should she pluck'd off, though indirect, Yet in conclusion is, I love this hurly all unwillingness will touch him; and all, ay, his trial. 
Ham. 
These eyes, because thou seest, ourselves have stomach, to't before; He hath committed; And then like a fever thee! ARVIRAGUS. 
I knew what thou hast thou make sick? Will conquer standing here remain Beyond all for it thinks your charge: you walk not deliver'd. 
O God, thou me for him. 
And take my lord. 
Nativity What fashion, and it open. 
SOOTHSAYER. 
PROSPERO. 
Bless thee, Jessica, I have, or red? Hor. 
Give me that any harm? Those whose charms By so stonish a swain's wearing; and too bad as snow, And she hath more To think in judgment I myself wrong, we know it is- but, good service, my pocket? Prince. 
He might have no softer bowels, Mo

At this point you might enjoy finding other sources of text on the internet and replacing the url at the top of this notebook. You can use song lyrics, news articles, etc. You just need to find files ending with '.txt'. Alternatively, you can copy and paste any block of text from any file into colab, assign it to the variable `text`, and then use wordlist=text.split() to get a list of words.

In [34]:
#This is the poem 'The Road Not Taken' by Robert frost
text= """Two roads diverged in a yellow wood,
And sorry I could not travel both
And be one traveler, long I stood
And looked down one as far as I could
To where it bent in the undergrowth;

Then took the other, as just as fair,
And having perhaps the better claim,
Because it was grassy and wanted wear;
Though as for that the passing there
Had worn them really about the same,

And both that morning equally lay
In leaves no step had trodden black.
Oh, I kept the first for another day!
Yet knowing how way leads on to way,
I doubted if I should ever come back.

I shall be telling this with a sigh
Somewhere ages and ages hence:
Two roads diverged in a wood, and I—
I took the one less traveled by,
And that has made all the difference."""
wordlist= text.split()
print(wordlist)

#the rest of the steps would continue as followed in the above blocks of code except with this new text chunk

['Two', 'roads', 'diverged', 'in', 'a', 'yellow', 'wood,', 'And', 'sorry', 'I', 'could', 'not', 'travel', 'both', 'And', 'be', 'one', 'traveler,', 'long', 'I', 'stood', 'And', 'looked', 'down', 'one', 'as', 'far', 'as', 'I', 'could', 'To', 'where', 'it', 'bent', 'in', 'the', 'undergrowth;', 'Then', 'took', 'the', 'other,', 'as', 'just', 'as', 'fair,', 'And', 'having', 'perhaps', 'the', 'better', 'claim,', 'Because', 'it', 'was', 'grassy', 'and', 'wanted', 'wear;', 'Though', 'as', 'for', 'that', 'the', 'passing', 'there', 'Had', 'worn', 'them', 'really', 'about', 'the', 'same,', 'And', 'both', 'that', 'morning', 'equally', 'lay', 'In', 'leaves', 'no', 'step', 'had', 'trodden', 'black.', 'Oh,', 'I', 'kept', 'the', 'first', 'for', 'another', 'day!', 'Yet', 'knowing', 'how', 'way', 'leads', 'on', 'to', 'way,', 'I', 'doubted', 'if', 'I', 'should', 'ever', 'come', 'back.', 'I', 'shall', 'be', 'telling', 'this', 'with', 'a', 'sigh', 'Somewhere', 'ages', 'and', 'ages', 'hence:', 'Two', 'roads'

## Bonus:

Recall the following recursive function that we used to define the Fibonacci numbers:

In [40]:
'''
def fibonacci(n):
  # Base cases: F(0) = 0, F(1) = 1
  if n == 0:
    return 0
  elif n == 1:
    return 1
  else:
    # Recursive case: F(n) = F(n-1) + F(n-2)
    return fibonacci(n - 1) + fibonacci(n - 2)
    '''


In [46]:
#modifying the fibonacci() function to accept an additional parameter count
def fibonacci(n, count=0):
    # Incrementing the count by 1 for each function call
    count += 1

    # Base cases: F(0) = 0, F(1) = 1
    #checking if n= 0 and 1 which are the base cases/ starting points of the sequence
    if n == 0:
        return 0, count
    elif n == 1:
        return 1, count
    else:
        # Recursive case: F(n) = F(n-1) + F(n-2)
        # If n is not 0 or 1, it recursively calls the fibonacci() function for n-1 and n-2 to calculate the Fibonacci numbers for those positions.
        #it then returns the fibonacci numbers along with the count
        fib_n_1, count = fibonacci(n - 1, count)
        fib_n_2, count = fibonacci(n - 2, count)
        return fib_n_1 + fib_n_2, count

# Calculate fibonacci(17) and get the count of function calls
fib_result, number_of_calls = fibonacci(17)
print("value returned by the function Fibonacci(17) =", fib_result)
print("Number of function calls:", number_of_calls)



value returned by the function Fibonacci(17) = 1597
Number of function calls: 5167


Here is the question.  If you run this line of code:  
`fibonacci(17)`  
how many times does the function `fibonacci()` get called?  
(The answer is not 1597, which is the value returned by the function.)